### PR TITLE
[WIP] Multi-stage (& multi-arch) dockerfile

### PR DESCRIPTION
**Pull request by sapk**
Implement multi-stage dockerfile. This will permit cleaner build with the ability to configure build params and maybe if FROM are multi-arch to have only one dockerfile. 

After that someone could simply do `docker build .` to build the image locally on every platform.  (Not yet ready)

### CODE REVIEW

**Comment by sapk** [comment]
This will need at least docker/moby@v17.05.0-ce-rc2 for multi-stage to work. So this PR is a little in advance. This will dependency is only for building multi-stage. The image resulting will be able to run even with older docker/moby version.

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: Why remove the Maintainer? Add yourself instead :trollface:

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: Maybe default to `sqlite`?

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: No need to `go get go-bindata` since it will be fetched by `make generate`

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: Why `pwd && ls -lah` ?

File: .dockerignore
Comment (bkcsoft) [inline .dockerignore]: this should def. be here if we're building it inside the container now

Comment (bkcsoft) [reply .dockerignore]: Heck, all there things you've added in here should still go into the first build...

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: `echo "git:$(dd if=/dev/urandom bs=24 count=1 | base64)" | chpasswd` (24 byte => base64 => 32 byte)

File: Dockerfile
Comment (bkcsoft) [inline Dockerfile]: We should still remove the cache...

File: Makefile
Comment (bkcsoft) [inline Makefile]: Is this stage still required?

**Review by bkcsoft — CHANGES_REQUESTED** [review]

**Comment by bkcsoft** [comment]
I added `status/blocked` since this is blocked on build-server getting Docker 17.05 **stable** 😉

File: Dockerfile
Comment (sapk) [reply Dockerfile]: Debug ^^

File: Dockerfile
Comment (sapk) [reply Dockerfile]: I can but with --no-cache it shouldn't be needed

File: Dockerfile
Comment (sapk) [reply Dockerfile]: It's only needed for the published image (last FROM)

File: Dockerfile
Comment (sapk) [reply Dockerfile]: I have re-tested and you are right. I remember that one time it ask for it but don't remember why.

Comment (sapk) [reply Dockerfile]: I think that it was because `git` wasn't installed so it couldn't be installed and failed after.

File: Dockerfile
Comment (sapk) [reply Dockerfile]: I haven't touch that.

File: Makefile
Comment (sapk) [reply Makefile]: I have separate it but still keep it if someone want to build binaries via docker.

File: Dockerfile
Comment (sapk) [reply Dockerfile]: And I use the LABEL instruction since MAINTAINER is depraced.

File: .dockerignore
Comment (sapk) [reply .dockerignore]: At startup docker "clone" local folder to build folder. This limit what is copied into workspace.

File: Dockerfile
Comment (bkcsoft) [reply Dockerfile]: Erm no, completely different things... Don't remove this line...

File: Dockerfile
Comment (bkcsoft) [reply Dockerfile]: right, it was more of a "since you're already screwing with it, might as well fix this" :trollface:

File: Dockerfile
Comment (sapk) [reply Dockerfile]: https://github.com/gliderlabs/docker-alpine/blob/master/docs/usage.md#disabling-cache It seem, it doesn't even need -U also.

Comment (sapk) [reply Dockerfile]: I have checked after apk --no-cache -> /var/cache/apk is empty.

**Comment by sapk** [comment]
To test against docker version in master : http://blog.alexellis.io/mutli-stage-docker-builds/#howcanitryitout

File: Makefile
Comment (bkcsoft) [inline Makefile]: no, `ARG TAGS="sqlite"` in the Dockerfile...

**Review by bkcsoft — CHANGES_REQUESTED** [review]

File: .dockerignore
Comment (sapk) [reply .dockerignore]: After re-thinking if we add all the repo (remove .dockerignore) we could be able to ask to compile a specific version (tag) at build stage via git.

**Comment by sapk** [comment]
At this stage, Dockerfile should be buildable on platform amd64, arm, aarch64 simply by tapping docker build .

I am figuring out how to build them in drone and do multi-arch image.

File: Dockerfile
Comment (sapk) [inline Dockerfile:35]: Reminder to myself : Check if this is usefull since this still run with root after.

**Comment by bkcsoft** [comment]
@sapk There's no way to build this on the drone server until it's upgraded to 17.05 🙂

File: Dockerfile
Comment (bkcsoft) [reply Dockerfile]: ooh, `apk --no-cache`, I thought you meant `docker build --no-cache` 😂 

Then it's fine 👍

**Review by bkcsoft — APPROVED** [review]

**Comment by sapk** [comment]
@bkcsoft we could by using the method explain in http://blog.alexellis.io/mutli-stage-docker-builds/#howcanitryitout but this would make use build last version of docker and start a container with this version in it and build this Dockerfile in it. Not really efficient.

We could also do the same without docker building step and re-use release tar.gz

**Comment by ulm0** [comment]
> I am figuring out how to build them in drone and do multi-arch image.

@sapk  I know it's not matter of this PR, but it'd be great if you make a HOW-TO on making a multi arch image, i'm interested in doing so as well.

**Comment by sapk** [comment]
@klud1 a multi-arch image is just a fake image linking to various image corresponding for arch (documented in files like https://github.com/sapk/dockerfiles/tree/master/alpine-multi) that you push with github.com/estesp/manifest-tool (you can see command-line example in https://github.com/sapk/dockerfiles/blob/master/.travis.yml that update the refs). 

And docker pull (like https://hub.docker.com/r/sapk/alpine-multi/tags/) it will choose the corresponding hash/image to pull for his running arch.
 
So you build a image for each arch and reference them in a .yml 

For the multi-arch part, I use https://github.com/portainer/docker-manifest as a reference and use multiarch/alpine for a good base to have image on most arch.

So with my alpine base multi-arch use as base, if you build this Dockerfile, it will/should result in a working on you current arch image.

**Comment by ulm0** [comment]
@sapk amazing, thank you so much.

**Comment by sapk** [comment]
I tested to do a` docker build` on an armhf platform of the multi-stage image with multi-arch sapk/alpine-multi as base and this work seamlessly. \o/

I think that it would be more clean to push this change under Dockerfile.multi and keep the olds ones around until the needed version for docker is ready if is is ok with you.

File: docker/manifest/gitea-1-0-0.yml
Comment (bkcsoft) [inline docker/manifest/gitea-1-0-0.yml]: What are all these manifests for?

**Review by bkcsoft — COMMENTED** [review]
otherwise LG-TM

File: docker/manifest/gitea-1-0-0.yml
Comment (sapk) [reply docker/manifest/gitea-1-0-0.yml]: For generating multi-arch image that reference the image to corresponding arch.

Comment (sapk) [reply docker/manifest/gitea-1-0-0.yml]: The only part missing is generation of image on arch. This could be done via qemu like https://github.com/multiarch/alpine does.

**Comment by appleboy** [comment]
We need to waiting for this PR https://github.com/drone-plugins/drone-docker/pull/124 for Drone to support multi-stage build.

**Comment by sapk** [comment]
Ok so for a little update. 

I remove olds Dockerfile.

The task `make docker` will build gitea/gitea for the platform you are running on (compatible with amd64, arm, arm64). (For this to fully work the multi-arch base gitea/gitea-base need to be set on docker hub by someone with rights. `DOCKER_MANIFEST=docker/manifest/base.yml make docker-multi-update-manifest` )

The tasks `make docker-multi-*` permit to cross-build and do multi-arch image. I don't think this is possible yet with drone-plugins/drone-docker.

**Comment by sapk** [comment]
It's also possible to choose the version gitea we want in the container via variable `GITEA_VERSION`. It accept any ref that is accepted by `git checkout `

**Comment by sapk** [comment]
I will break this into multiple PR to ease the process of validation.
- Unify dockerfile with one multi-arch as base. (permit to build the same dockerfile on every platform)
- Multi-stage (permit to have a stage build to limit image size)
- Multi-arch final image (permit to download from docker hub the corresponding image of your running arch with the same name for every platform)

**Pull request closed without merging**
